pkg/reconciler/workload/synctargetexports/synctaargetcompatible_reconcile.go

```diff
@@ -0,0 +1,133 @@
+type apiCompatibleReconciler struct {
+	getAPIExport           func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExport, error)
+	getResourceSchema      func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIResourceSchema, error)
+	listAPIResourceImports func(clusterName logicalcluster.Name) ([]*apiresourcev1alpha1.APIResourceImport, error)
+}
+
+func (e *apiCompatibleReconciler) reconcile(ctx context.Context, syncTarget *workloadv1alpha1.SyncTarget) (*workloadv1alpha1.SyncTarget, reconcileStatus, error) {
+	exportKeys := getExportKeys(syncTarget)
+
+	var errs []error
+	schemaMap := map[schema.GroupVersionResource]*apiextensionsv1.JSONSchemaProps{}
+
+	// Get json schema from all related resource schemas
+	for _, exportKey := range exportKeys {
+		exportCluster, name := clusters.SplitClusterAwareKey(exportKey)
+		export, err := e.getAPIExport(exportCluster, name)
+		if apierrors.IsNotFound(err) {
+			continue
+		}
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		for _, schemaName := range export.Spec.LatestResourceSchemas {
+			resourceSchema, err := e.getResourceSchema(exportCluster, schemaName)
+			if apierrors.IsNotFound(err) {
+				continue
+			}
+			if err != nil {
+				errs = append(errs, err)
+			}
+
+			for _, v := range resourceSchema.Spec.Versions {
+				jsonSchema, err := v.GetSchema()
+				if err != nil {
+					errs = append(errs, err)
+					continue
+				}
+				schemaMap[schema.GroupVersionResource{
+					Group:    resourceSchema.Spec.Group,
+					Resource: resourceSchema.Spec.Names.Plural,
+					Version:  v.Name,
+				}] = jsonSchema
+			}
+		}
+	}
+
+	lcluster := logicalcluster.From(syncTarget)
+	apiImportMap := map[schema.GroupVersionResource]*apiextensionsv1.JSONSchemaProps{}
+	apiImports, err := e.listAPIResourceImports(lcluster)
+	if err != nil {
+		return syncTarget, reconcileStatusStop, err
+	}
+
+	for _, apiImport := range apiImports {
+		jsonSchema, err := apiImport.Spec.GetSchema()
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		apiImportMap[schema.GroupVersionResource{
+			Group:    apiImport.Spec.GroupVersion.Group,
+			Version:  apiImport.Spec.GroupVersion.Version,
+			Resource: apiImport.Spec.Plural,
+		}] = jsonSchema
+	}
+
+	for i, syncedRsesource := range syncTarget.Status.SyncedResources {
+		for _, v := range syncedRsesource.Versions {
+			gvr := schema.GroupVersionResource{Group: syncedRsesource.Group, Resource: syncedRsesource.Resource, Version: v}
+			upstreamSchema, ok := schemaMap[gvr]
+			if !ok {
+				syncTarget.Status.SyncedResources[i].State = workloadv1alpha1.ResourceSchemaPendingState
+				continue
+			}
+
+			downStreamSchema, ok := apiImportMap[gvr]
+			if !ok {
+				syncTarget.Status.SyncedResources[i].State = workloadv1alpha1.ResourceSchemaIncomptibleState
+				continue
+			}
+
+			_, err := schemacompat.EnsureStructuralSchemaCompatibility(
```

Comment:
am unsure we really want this reconciler today. We will rework this API import logic with the API management topic. We can keep the reconciler for now, but should be prepared that this area will change quite a bit. If the reconciler makes trouble, wouldn't have a problem to disable it.